ir/src/main/scala/org/scalajs/ir/Serializers.scala

```diff
@@ -1168,18 +1179,78 @@ object Serializers {
     }
 
     def readOptHash(): Option[TreeHash] = {
-      if (input.readBoolean()) {
+      if (readBoolean()) {
         val hash = new Array[Byte](20)
-        input.readFully(hash)
+        buf.get(hash)
         Some(new TreeHash(hash))
-      } else None
+      } else {
+        None
+      }
     }
 
     def readString(): String = {
-      strings(input.readInt())
+      strings(readInt())
     }
 
     def readStrings(): List[String] =
-      List.fill(input.readInt())(readString())
+      List.fill(readInt())(readString())
+
+    private def readBoolean() = buf.get() != 0
+    private def readByte() = buf.get()
+    private def readChar() = buf.getChar()
+    private def readShort() = buf.getShort()
+    private def readInt() = buf.getInt()
+    private def readLong() = buf.getLong()
+    private def readFloat() = buf.getFloat()
+    private def readDouble() = buf.getDouble()
+
+    private def readUTF(): String = {
+      // DataInput.readUTF for buffers.
+
+      val length = buf.getShort() & 0xffff // unsigned
+      var res = ""
+      var i = 0
+
+      def badFormat(msg: String) = throw new UTFDataFormatException(msg)
+
+      while (i < length) {
+        val a = buf.get()
+
+        i += 1
+
+        val char = {
+          if ((a & 0x80) == 0x00) { // 0xxxxxxx
+            a.toChar
+          } else if ((a & 0xE0) == 0xC0 && i < length) { // 110xxxxx
+            val b = buf.get()
+            i += 1
+
+            if ((b & 0xC0) != 0x80) // 10xxxxxx
+              badFormat(f"Expected 2 bytes, found: $b%#02x (init: $a%#02x)")
```

Comment:
Oh, I almost missed this: we need to avoid using the `f` interpolator in the `ir/` project, otherwise dotty cannot compile the sources as part of its own compiler (since it's a macro, and when dotty compiles its own compiler, macros don't work). Explicitly use `"...".format(...)` instead.

See also https://github.com/scala-js/scala-js/pull/3597

(there are a few other `f"..."` in this method)